client/web/src/components/externalServices/AddExternalServicesPage.tsx

```diff
@@ -120,102 +124,162 @@ export const AddExternalServicesPage: FC<AddExternalServicesPageProps> = ({
+            code data to Sourcegraph.com or any other destination. Your code is kept private on this installation.
+        </Text>
+        <H3>This Sourcegraph installation will access your code host by:</H3>
+        <ul>
+            <li>
+                Periodically fetching a list of repositories to ensure new, removed, and renamed repositories are
+                accessible on Sourcegraph.
+            </li>
+            <li>Cloning the repositories you specify to create a local cache.</li>
+            <li>Periodically pulling cloned repositories to ensure search results are current.</li>
+            <li>
+                Fetching{' '}
+                <Link to="/help/admin/permissions" target="_blank" rel="noopener noreferrer">
+                    user repository access permissions
+                </Link>
+                , if you have enabled this feature.
+            </li>
+            <li>
+                Opening pull requests and syncing their metadata as part of{' '}
+                <Link to="/help/batch_changes" target="_blank" rel="noopener noreferrer">
+                    batch changes
+                </Link>
+                , if you have enabled this feature.
+            </li>
+        </ul>
+        <div className="d-flex justify-content-end">
+            <Button variant="secondary" onClick={dismissPrivacyWarning}>
+                Do not show this again
+            </Button>
+        </div>
+    </Alert>
+)
+
+interface ExternalServicesGroup {
+    services: AddExternalServiceOptionsWithID[]
+    icon: React.ComponentType<{ className?: string }>
+    description: string
+    renderServiceIcon: boolean
+}
+
+type GroupedServiceDisplayName = 'GitHub' | 'GitLab' | 'Bitbucket' | 'Other code hosts'
+
+const computeExternalServicesGroup = (
+    services: Record<string, AddExternalServiceOptions>,
+    allowedCodeHosts: AddExternalServiceOptions[] | null
+): Record<GroupedServiceDisplayName, ExternalServicesGroup> => {
+    const groupedServices: Record<GroupedServiceDisplayName, ExternalServicesGroup> = {
+        GitHub: {
+            services: [],
+            icon: GithubIcon,
+            description: 'Connect with repositories on GitHub',
+            renderServiceIcon: false,
+        },
+        GitLab: {
+            services: [],
+            icon: GitLabIcon,
+            description: 'Connect with repositories on GitLab',
+            renderServiceIcon: false,
+        },
+        Bitbucket: {
+            services: [],
+            icon: BitbucketIcon,
+            description: 'Connect with repositories on Bitbucket',
+            renderServiceIcon: false,
+        },
+        'Other code hosts': { services: [], icon: GitIcon, description: '', renderServiceIcon: true },
+    }
+
+    for (const [serviceID, service] of Object.entries(services)) {
+        const isDisabled = Boolean(allowedCodeHosts && !allowedCodeHosts.includes(service))
```

Comment:
If using Set as suggested above, this can be simplified to
```suggestion
        const isDisabled = !allowedCodeHosts?.has(service)
```

The `!` operator already converts the statement to boolean. Calling `has` on an empty set is practically a noop. And if `allowedCodeHosts` is not defined, `!undefined === true`